packages/flutter/lib/src/material/button_bar.dart

```diff
@@ -172,3 +181,183 @@ class ButtonBar extends StatelessWidget {
+      return super.constraints;
+    return super.constraints.copyWith(maxWidth: double.infinity);
+  }
+
+  @override
+  void performLayout() {
+    // Set check layout width to false in reload or update cases.
+    _hasCheckedLayoutWidth = false;
+
+    // Perform layout to ensure that button bar knows how wide it would
+    // ideally want to be.
+    super.performLayout();
+    _hasCheckedLayoutWidth = true;
+
+    // If the button bar is constrained by width and it overflows, set the
+    // buttons to align vertically. Otherwise, lay out the button bar
+    // horizontally.
+    if (size.width > constraints.maxWidth) {
+      RenderBox child = firstChild;
+      double currentHeight = 0.0;
+      while (child != null) {
+        final FlexParentData childParentData = child.parentData;
+        assert(child.parentData == childParentData);
+
+        // Lay out the child with the button bar's original constraints, but
+        // with minimum width set to zero.
+        child.layout(constraints.copyWith(minWidth: 0.0), parentUsesSize: true);
+
+        // Set the cross axis alignment for the column to match the main axis
+        // alignment for a row. For [MainAxisAligment.spaceAround],
+        // [MainAxisAligment.spaceBetween] and [MainAxisAlignment.spaceEvenly]
+        // cases, use [MainAxisAligmnent.start].
+        switch (textDirection) {
+          case TextDirection.ltr:
+            switch (mainAxisAlignment) {
+              case MainAxisAlignment.center:
+                final double midpoint = (constraints.maxWidth - child.size.width) / 2.0;
+                childParentData.offset = Offset(midpoint, currentHeight);
+                break;
+              case MainAxisAlignment.end:
+                childParentData.offset = Offset(constraints.maxWidth - child.size.width, currentHeight);
+                break;
+              default:
+                childParentData.offset = Offset(0, currentHeight);
+                break;
+            }
+            break;
+          case TextDirection.rtl:
+            switch (mainAxisAlignment) {
+              case MainAxisAlignment.center:
+                final double midpoint = constraints.maxWidth / 2.0 - child.size.width / 2.0;
+                childParentData.offset = Offset(midpoint, currentHeight);
+                break;
+              case MainAxisAlignment.end:
+                childParentData.offset = Offset(0, currentHeight);
+                break;
+              default:
+                childParentData.offset = Offset(constraints.maxWidth - child.size.width, currentHeight);
+                break;
+            }
+            break;
+          default:
+            assert(true, 'Text direction is null when it should be set.');
+            break;
+        }
+        currentHeight += child.size.height;
+        child = childParentData.nextSibling;
+      }
+      size = constraints.constrain(Size(constraints.maxWidth, currentHeight));
+    } else {
```

Comment:
Is it really necessary to performLayout() again, if the row fit? If so, this would be a good place to explain why.

It might be clearer if this clause was first.

Reply:
It is, since the original `performLayout()` call uses infinity as the maximum width constraint. The second layout is performed to ensure that we use the original maximum width. I'll add a comment to explain this and have it as the first clause.